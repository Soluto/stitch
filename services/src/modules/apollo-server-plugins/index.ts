import { PluginDefinition } from 'apollo-server-core';
import { transformApolloServerPlugins as applyPluginsForApolloServerPlugins } from '../plugins';
import { createBasicPolicyPlugin } from './base-policy';
import { createIntrospectionQueryPolicyPlugin } from './introspection-query-policy';
import { createLoggingPlugin } from './logging';
import { createMetricsPlugin } from './metrics';

export default function getPlugins(): PluginDefinition[] {
  const baseApolloServerPlugins = [
    createBasicPolicyPlugin,
    createLoggingPlugin,
    createMetricsPlugin,
    createIntrospectionQueryPolicyPlugin,
  ];
  const plugins = applyPluginsForApolloServerPlugins(baseApolloServerPlugins);
  return plugins;
}
