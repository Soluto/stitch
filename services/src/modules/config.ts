import * as path from 'path';
import * as envVar from 'env-var';
import { LoggerOptions, LevelWithSilent } from 'pino';
// import { PlaygroundConfig } from 'apollo-server-core';
import { AuthenticationConfig } from './authentication/types';
import { CorsConfiguration } from './cors';

const envVarExt = envVar.from(process.env, {
  asSet: (value: string) => new Set(value.split(',')),
});

// General
export const nodeEnv = envVarExt.get('NODE_ENV').default('development').asString();

export const httpPort = envVarExt.get('PORT').default('8080').asIntPositive();
export const keepAliveTimeout = envVarExt.get('KEEP_ALIVE_TIMEOUT').asIntPositive();

// Logging
export type ChildLoggersLevels = Record<string, LevelWithSilent>;
export const logLevel = envVarExt.get('LOG_LEVEL').default('WARN').asString();
export const childLoggerLevels = envVarExt.get('CHILD_LOGGERS_LEVELS').default({}).asJsonObject() as ChildLoggersLevels;
export const loggerConfiguration = envVar.get('LOGGER_CONFIGURATION').default({}).asJsonObject() as LoggerOptions;

// Repositories
export const useS3ResourceRepository = envVarExt.get('USE_S3_RESOURCE_REPOSITORY').default('false').asBoolStrict();
export const useFileSystemResourceRepository = envVarExt
  .get('USE_FS_RESOURCE_REPOSITORY')
  .default('true')
  .asBoolStrict();

// Resources
export const resourceUpdateInterval = envVarExt.get('RESOURCE_UPDATE_INTERVAL').default('60000').asIntPositive();

// GraphQL configuration
export const enableGraphQLTracing = envVarExt.get('GRAPHQL_TRACING').default('true').asBool();
// export const enableGraphQLPlayground = envVarExt
//   .get('GRAPHQL_PLAYGROUND')
//   .default('true')
//   .asJsonObject() as PlaygroundConfig;
export const enableGraphQLIntrospection = envVarExt.get('GRAPHQL_INTROSPECTION').default('true').asBoolStrict();

// Policies
export const tmpPoliciesDir = envVarExt
  .get('TMP_POLICIES_DIR')
  .default(path.resolve(process.cwd(), 'tmp/policies'))
  .asString();

export const ignorePolicies =
  envVar.get('NODE_ENV').required().asString() !== 'production' &&
  envVar.get('IGNORE_POLICIES').default('false').asBoolStrict();

// Authentication
const defaultAuthenticationConfig: AuthenticationConfig = {
  anonymous: {
    publicPaths: ['/metrics', '/.well-known/apollo/server-health', '/graphql', '/status'],
  },
};
export const authenticationConfig =
  (envVarExt.get('AUTHENTICATION_CONFIGURATION').asJsonObject() as AuthenticationConfig) ?? defaultAuthenticationConfig;

// Plugins
export const pluginsDir = envVar.get('PLUGINS_DIR').asString();
export const pluginsConfig = envVar.get('PLUGINS_CONFIGURATION').asJsonObject() as Record<string, unknown> | undefined;

// TODO:Check the list of apollo federation directives when upgrading apollo version
const defaultKnownApolloDirectives = 'key,extends,external,requires,provides';
export const knownApolloDirectives = envVarExt
  .get('KNOWN_APOLLO_DIRECTIVES')
  .default(defaultKnownApolloDirectives)
  .asSet();

export const corsConfiguration = envVar.get('CORS_CONFIGURATION').default({}).asJsonObject() as CorsConfiguration;

export const requestDurationPromHistogramBuckets: number[] = envVarExt
  .get('REQUEST_DURATION_PROM_HISTOGRAM_BUCKETS')
  .default('[0.02, 0.1, 0.5, 2, 10]')
  .asJsonArray();

export const resolverDurationPromHistogramBuckets: number[] = envVarExt
  .get('RESOLVER_DURATION_PROM_HISTOGRAM_BUCKETS')
  .default('[0.02, 0.1, 0.5, 2, 10]')
  .asJsonArray();
