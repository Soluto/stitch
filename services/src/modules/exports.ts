// import { GraphQLObjectType, GraphQLResolveInfo, GraphQLField, GraphQLInterfaceType } from 'graphql';
// import { GraphQLExtension } from 'apollo-server-core';
// import { RequestContext } from './context';

// function isObject(val: unknown): val is Record<string, unknown> {
//   return new Object(val) === val;
// }

// export class ExportTrackingExtension implements GraphQLExtension<RequestContext> {
//   objectToParentMap = new Map<unknown, { parent: unknown; parentType: GraphQLObjectType | GraphQLInterfaceType }>();

//   requestDidStart(options: { context: RequestContext }) {
//     options.context.exports = this;
//   }

//   willResolveField(
//     source: unknown,
//     _args: unknown,
//     _context: RequestContext,
//     info: GraphQLResolveInfo
//   ): ((error: Error | null, result?: unknown) => void) | void {
//     return (error, result) => {
//       if (error || result === null || typeof result === 'undefined') {
//         return;
//       }

//       // If the result is not an object or array, there's never a need to look it up because it won't have children
//       // Therefore, we don't need to set it up in the parentMap
//       // This means parentMap can also be a weakMap if we want it to be
//       if (Array.isArray(result)) {
//         for (const element of result) {
//           if (isObject(element)) {
//             this.objectToParentMap.set(element, { parent: source, parentType: info.parentType });
//           }
//         }
//       } else if (isObject(result)) {
//         this.objectToParentMap.set(result, { parent: source, parentType: info.parentType });
//       }
//     };
//   }

//   resolve(
//     objectType: GraphQLObjectType | GraphQLInterfaceType,
//     objectValue: Record<string, unknown>,
//     exportKey: string
//   ): unknown {
//     const parentExportToFieldMap = exportMap.get(objectType);

//     if (typeof parentExportToFieldMap !== 'undefined') {
//       const exportingFieldName = parentExportToFieldMap.get(exportKey);

//       if (typeof exportingFieldName !== 'undefined') {
//         return objectValue[exportingFieldName];
//       }
//     }

//     const parentDetails = this.objectToParentMap.get(objectValue);

//     if (typeof parentDetails === 'undefined') {
//       return;
//     }

//     return this.resolve(parentDetails.parentType, parentDetails.parent as Record<string, unknown>, exportKey);
//   }
// }

// const exportMap = new WeakMap<GraphQLObjectType | GraphQLInterfaceType, Map<string, string>>();
// export function markExport(
//   object: GraphQLObjectType | GraphQLInterfaceType,
//   field: GraphQLField<unknown, RequestContext>,
//   exportKey: string
// ) {
//   let exportToFieldMap = exportMap.get(object);
//   if (typeof exportToFieldMap === 'undefined') {
//     exportToFieldMap = new Map<string, string>();
//     exportMap.set(object, exportToFieldMap);
//   }

//   exportToFieldMap.set(exportKey, field.name);
// }

// declare module './context' {
//   interface RequestContext {
//     exports: Pick<ExportTrackingExtension, 'resolve'>;
//   }
// }

// type ExportProxyTarget = {
//   exports: ExportTrackingExtension;
//   objectType: GraphQLObjectType | GraphQLInterfaceType;
//   objectValue: Record<string, unknown>;
// };

// const handler = {
//   get: (target: ExportProxyTarget, property: string) => {
//     const result = target.exports.resolve(target.objectType, target.objectValue, property);
//     return result;
//   },
// };

// export function getExportsProxy(
//   exports: Pick<ExportTrackingExtension, 'resolve'>,
//   objectType: GraphQLObjectType | GraphQLInterfaceType,
//   objectValue: Record<string, unknown>
// ) {
//   const target = {
//     exports,
//     objectType,
//     objectValue,
//   };

//   return new Proxy(target, handler);
// }
