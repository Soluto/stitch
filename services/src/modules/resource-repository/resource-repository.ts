import * as path from 'path';
import * as pLimit from 'p-limit';
import * as _ from 'lodash';
import { Storage, listFilesItem } from '../storage';
import { getWasmPolicy } from '../directives/policy/opa';
import buildResourcesMetadata from './resource-metadata';
import {
  FetchLatestResult,
  ResourceGroup,
  IResourceRepository,
  PolicyAttachments,
  UpdateOptions,
  ResourcesMetadata,
} from '.';

export class ResourceRepository implements IResourceRepository {
  protected resourceGroup?: { lastModified?: Date; resources: ResourceGroup };
  protected policyAttachments: { attachments: PolicyAttachments; refreshedAt?: Date } = { attachments: {} };

  constructor(
    protected storage: Storage,
    protected resourceFilePath: string,
    protected policyAttachmentsFolderPath: string,
    protected isRegistry = false,
    protected registryResourceFilePath?: string
  ) {}

  async fetchLatest(): Promise<FetchLatestResult> {
    const resourceFilePath = this.isRegistry ? this.registryResourceFilePath! : this.resourceFilePath;
    const stats = await this.storage.fileStats(resourceFilePath);

    if (this.resourceGroup?.lastModified?.getTime() === stats.lastModified.getTime()) {
      await this.refreshPolicyAttachments();
      this.resourceGroup.resources.policyAttachments = { ...this.policyAttachments.attachments };

      return { isNew: false, resourceGroup: this.resourceGroup.resources };
    }

    const { content } = await this.storage.readFile(resourceFilePath, { asString: true });
    const resources = JSON.parse(content) as ResourceGroup;
    this.resourceGroup = { lastModified: stats.lastModified, resources };

    await this.refreshPolicyAttachments();
    this.resourceGroup.resources.policyAttachments = { ...this.policyAttachments.attachments };

    return { isNew: true, resourceGroup: resources };
  }

  async fetchMetadata(): Promise<ResourcesMetadata> {
    const resourceFilePath = this.isRegistry ? this.registryResourceFilePath! : this.resourceFilePath;
    const resourceMetadataFilePath = this.getMetadataFilePath(resourceFilePath);

    const { content } = await this.storage.readFile(resourceMetadataFilePath, { asString: true });
    const metadata = JSON.parse(content) as ResourcesMetadata;

    return metadata;
  }

  async update(resources: ResourceGroup, options?: UpdateOptions): Promise<void> {
    const defaultOptions = { registry: false };
    const { registry } = { ...defaultOptions, ...options };
    const resourceFilePath = registry ? this.registryResourceFilePath! : this.resourceFilePath;
    const resourceFileContent = JSON.stringify(resources);

    const resourceMetadataFilePath = this.getMetadataFilePath(resourceFilePath);
    const resourceMetadata = await buildResourcesMetadata(resources, resourceFileContent, options);

    await this.storage.writeFile(resourceFilePath, resourceFileContent);
    await this.storage.writeFile(resourceMetadataFilePath, resourceMetadata);
  }

  async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
    const filePath = path.join(this.policyAttachmentsFolderPath, filename);
    await this.storage.writeFile(filePath, content);
  }

  async deletePolicyAttachment(filename: string): Promise<void> {
    const filePath = path.join(this.policyAttachmentsFolderPath, filename);
    await this.storage.deleteFile(filePath);
  }

  protected async refreshPolicyAttachments() {
    if (this.isRegistry) return;
    const newRefreshedAt = new Date();

    const allAttachments = await this.listPolicyAttachments();
    const attachmentsToRefresh = allAttachments.filter(a => this.shouldRefreshPolicyAttachment(a)).map(a => a.filePath);

    if (attachmentsToRefresh.length > 0) {
      const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);

      for (const { filename, content } of newAttachments) {
        this.policyAttachments.attachments[filename] = await getWasmPolicy(content);
      }
    }

    const allNewFilePaths = allAttachments.map(a => path.basename(a.filePath));
    const allCurrentFilePaths = Object.keys(this.policyAttachments.attachments);
    const deletedAttachments = _.differenceWith(allCurrentFilePaths, allNewFilePaths, _.isEqual);
    for (const deletedAttachment of deletedAttachments) {
      delete this.policyAttachments.attachments[deletedAttachment];
    }

    this.policyAttachments.refreshedAt = newRefreshedAt;
  }

  async listPolicyAttachments() {
    return this.storage.listFiles(this.policyAttachmentsFolderPath);
  }

  protected shouldRefreshPolicyAttachment({ filePath, lastModified }: listFilesItem): boolean {
    const filename = path.basename(filePath);

    if (!this.policyAttachments.attachments[filename]) return true;
    if (!this.policyAttachments.refreshedAt) return true;

    return lastModified > this.policyAttachments.refreshedAt;
  }

  protected async getPolicyAttachments(filePaths: string[]): Promise<{ filename: string; content: Buffer }[]> {
    const limit = pLimit(10);

    return Promise.all(
      filePaths.map(filePath => {
        return limit(async () => {
          const { content } = await this.storage.readFile(filePath);
          const filename = path.basename(filePath);

          return { filename, content };
        });
      })
    );
  }

  protected getMetadataFilePath(resourceFilePath: string) {
    return `${resourceFilePath.split('.').slice(0, -1).join('.')}-metadata.json`;
  }
}
