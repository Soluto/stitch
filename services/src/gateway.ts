import fastify, { FastifyInstance } from 'fastify';
import corsPlugin from '@fastify/cors';
import * as fastifyMetrics from 'fastify-metrics';
import jwtPlugin from '@fastify/jwt';
import authPlugin from '@fastify/auth';
import * as config from './modules/config';
import logger, { createChildLogger } from './modules/logger';
import { handleSignals, handleUncaughtErrors } from './modules/shutdown-handler';
import createStitchGateway from './modules/apollo-server';
import { getSecret, anonymousPlugin, jwtDecoderPlugin, authStrategies } from './modules/authentication';
import { loadPlugins } from './modules/plugins';
import { initializeMetrics } from './modules/apollo-server-plugins/metrics';
// import { contentTypeFilterMiddleware } from './modules/fastify-middlewares';

const sLogger = createChildLogger(logger, 'http-server');

export async function createServer() {
  sLogger.info('Stitch gateway booting up...');

  await loadPlugins();

  const { server, updateSchema } = await createStitchGateway({
    // tracing: config.enableGraphQLTracing,
    // playground: config.enableGraphQLPlayground,
    introspection: config.enableGraphQLIntrospection,
  });
  const app = fastify();
  // appWithMiddlewares.use(contentTypeFilterMiddleware); FIXME: WHAT TO DO WITH THIS?

  await app
    .register(corsPlugin, config.corsConfiguration)
    .register(authPlugin)
    .register(jwtPlugin, {
      secret: getSecret,
      verify: {
        algorithms: ['RS256'],
      },
    })
    .register(jwtDecoderPlugin)
    .register(anonymousPlugin)
    .register(fastifyMetrics, {
      endpoint: '/metrics',
    })
    .register(server.createHandler({ path: '/graphql', cors: false }));

  app.after(() => {
    app.addHook('onRequest', app.auth(authStrategies));
    app.addHook('onClose', () => server.stop());
  });

  app.get('/health', async (_, reply) => {
    await reply.status(200).send('OK');
  });

  app.post('/updateSchema', async (_, reply) => {
    await updateSchema();
    await reply.status(200).send('OK');
  });

  app.server.keepAliveTimeout = config.keepAliveTimeout || app.server.keepAliveTimeout;

  sLogger.info('Stitch gateway is ready to start');
  return app;
}

async function runServer(app: FastifyInstance) {
  const address = await app.listen(config.httpPort, '0.0.0.0');
  initializeMetrics(app.metrics.client);
  sLogger.info({ address }, 'Stitch gateway started');

  handleSignals(() => app.close(() => sLogger.info('Stitch gateway stopped successfully')));
  handleUncaughtErrors();
}

// Only run when file is being executed, not when being imported
if (require.main === module) {
  // eslint-disable-next-line promise/catch-or-return
  void createServer().then(runServer);
}
