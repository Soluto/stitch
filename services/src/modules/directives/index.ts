import { concatAST, GraphQLSchema } from 'graphql';
import { ResourceGroup } from '../resource-repository';
import { sdl as stubSdl, directiveSchemaTransformer as stubDirectiveSchemaTransformer } from './stub';
import {
  sdl as localResolverSdl,
  directiveSchemaTransformer as localResolverDirectiveSchemaTransformer,
} from './local-resolver';
import {
  sdl as errorHandlerSdl,
  directiveSchemaTransformer as errorHandlerDirectiveSchemaTransformer,
} from './error-handler';
import {
  sdl as enumResolverSdl,
  directiveSchemaTransformer as enumResolverDirectiveSchemaTransformer,
} from './enum-resolver';
import { sdl as enumValueSdl, directiveSchemaTransformer as EnumValueDirective } from './enum-value';
import { sdl as restSdl, directiveSchemaTransformer as restDirectiveSchemaTransformer } from './rest';
import { sdl as gqlSdl, directiveSchemaTransformer as gqlDirectiveSchemaTransformer } from './gql';
// import { sdl as exportSdl, ExportDirective } from './export';
import { sdl as selectSdl, directiveSchemaTransformer as selectDirectiveSchemaTransformer } from './select';
import {
  policyBaseSdl,
  policySdl,
  policiesSdl,
  policyFieldSdl,
  policyQuerySdl,
  policyDirectiveSchemaTransformer,
  policiesDirectiveSchemaTransformer,
  policyQueryDirectiveSchemaTransformer,
  policyScalarResolvers,
} from './policy';

export const directiveMap: {
  [visitorName: string]: (schema: GraphQLSchema, context?: { resourceGroup: ResourceGroup }) => GraphQLSchema;
} = {
  stub: stubDirectiveSchemaTransformer,
  rest: restDirectiveSchemaTransformer,
  gql: gqlDirectiveSchemaTransformer,
  // export: ExportDirective,
  select: selectDirectiveSchemaTransformer,
  policy: policyDirectiveSchemaTransformer,
  policies: policiesDirectiveSchemaTransformer,
  policyQuery: policyQueryDirectiveSchemaTransformer,
  localResolver: localResolverDirectiveSchemaTransformer,
  errorHandler: errorHandlerDirectiveSchemaTransformer,
  enumResolver: enumResolverDirectiveSchemaTransformer,
  enumValue: EnumValueDirective,
};

export const sdl = concatAST([
  stubSdl,
  restSdl,
  gqlSdl,
  // exportSdl,
  selectSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  localResolverSdl,
  errorHandlerSdl,
  enumResolverSdl,
  enumValueSdl,
]);

export { policyScalarResolvers, policyBaseSdl, policyFieldSdl };
