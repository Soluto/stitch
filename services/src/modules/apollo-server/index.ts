import { ApolloServer, Config } from 'apollo-server-fastify';
import { ignorePolicies } from '../config';
import { RESTDirectiveDataSource } from '../directives/rest';
// import { ExportTrackingExtension } from '../exports';
import getPlugins from '../apollo-server-plugins';
import createGraphQLService from './create-graphql-service';

export interface StitchGatewayService {
  server: ApolloServer;
  updateSchema: () => Promise<void>;
}

export default async function createStitchGateway(apolloConfig?: Config): Promise<StitchGatewayService> {
  const gateway = createGraphQLService();
  const server = new ApolloServer({
    ...apolloConfig,
    gateway,
    // extensions: [() => new ExportTrackingExtension()], // TODO: Replace extensions with a plugin
    plugins: getPlugins(),
    context({ request }) {
      const ctx = {
        request,
        ignorePolicies,
      };
      return ctx;
    },
    dataSources() {
      return {
        rest: new RESTDirectiveDataSource(),
      };
    },
  });
  await server.start();

  return {
    server,
    updateSchema: async () => {
      await gateway.updateSchema();
    },
  };
}

export { default as createGatewaySchema } from './create-gateway-schema';
