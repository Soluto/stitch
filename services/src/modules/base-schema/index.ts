import { IResolvers } from '@graphql-tools/utils';
import { concatAST, DocumentNode, GraphQLSchema } from 'graphql';
import { policyBaseSdl, sdl as directivesSdl, directiveMap } from '../directives';
import { transformBaseSchema as applyPlugins } from '../plugins';
import { ResourceGroup } from '../resource-repository';
import { sdl as scalarsSdl, resolvers as scalarResolvers } from './scalars';

export type BaseSchema = {
  typeDefs: DocumentNode;
  resolvers: IResolvers;
  directives: Record<string, (schema: GraphQLSchema, context?: { resourceGroup: ResourceGroup }) => GraphQLSchema>;
};

export default async (): Promise<BaseSchema> => {
  const baseSchema = {
    typeDefs: concatAST([scalarsSdl, directivesSdl, policyBaseSdl]),
    resolvers: scalarResolvers,
    directives: directiveMap,
  };
  const baseSchemaWithPlugins = await applyPlugins(baseSchema);
  return baseSchemaWithPlugins;
};
