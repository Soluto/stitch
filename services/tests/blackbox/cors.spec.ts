import * as fs from 'fs/promises';
import { FastifyInstance } from 'fastify';
import { createServer as createGateway } from '../../src/gateway';
import { ResourceGroup } from '../../src/modules/resource-repository';

describe('CORS', () => {
  let app: FastifyInstance;

  beforeAll(async () => {
    const resources: ResourceGroup = {
      schemas: [],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [],
    };

    await fs.writeFile(process.env.FS_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));

    app = await createGateway();
  });

  afterAll(async () => {
    await app.close();
    await fs.unlink(process.env.FS_RESOURCE_REPOSITORY_PATH!);
  });

  test('Preflight request from valid origin', async () => {
    const response = await app.inject({
      method: 'OPTIONS',
      url: '/graphql',
      headers: {
        origin: 'http://localhost',
        'access-control-request-method': 'POST',
        'access-control-request-headers': 'x-api-client,authorization',
      },
    });

    expect(response.statusCode).toEqual(204);
    const headers = Object.fromEntries(
      Object.entries(response.headers).filter(([k]) => k.startsWith('access-control-allow'))
    );
    expect(headers).toMatchSnapshot();
  });

  test('Preflight request from invalid origin', async () => {
    const response = await app.inject({
      method: 'OPTIONS',
      url: '/graphql',
      headers: {
        origin: 'http://unknown-host',
        'access-control-request-method': 'PUT',
        'access-control-request-headers': 'x-api-client,authorization,unknown-header',
      },
    });

    expect(response.statusCode).toEqual(204);
    const headers = Object.fromEntries(
      Object.entries(response.headers).filter(([k]) => k.startsWith('access-control-allow'))
    );
    expect(headers).toMatchSnapshot();
  });
});
