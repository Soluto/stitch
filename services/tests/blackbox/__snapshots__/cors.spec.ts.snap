// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CORS Preflight request from valid origin 1`] = `
Object {
  "access-control-allow-headers": "x-api-client, authorization",
  "access-control-allow-methods": "GET, POST",
  "access-control-allow-origin": "http://localhost",
}
`;

exports[`CORS Preflight request from invalid origin 1`] = `
Object {
  "access-control-allow-headers": "x-api-client, authorization",
  "access-control-allow-methods": "GET, POST",
  "access-control-allow-origin": false,
}
`;
